### PR TITLE
Add contract instances filter support with a get param (fix #3971).

### LINKED ISSUE #3971: Query completed contract instances via PAB api
## Area
[] *Plutus Application Framework* Related to the Plutus application backend (PAB), emulator, Plutus libraries

## Describe the feature you'd like

At the moment we can query active contract instances via
curl -s http://localhost:9080/api/contract/instances

Sometimes it is useful to get data from the completed instances for historic data
At the moment I can query completed contract state via, 
```
curl -H "Content-Type: application/json" \
  --request GET \
  http://localhost:9080/api/contract/instance/$INSTANCE_ID/status
```
only there is no endpoint to query completed contract instance id, so I need to store $INSTANCE_ID offchain

Could we add endpoint to query completed contract instance id?
Is it safe to rely that the completed contract instance state will be stored in the PAB storage for the long time, or it is get purged eventually?

### PR DESCRIPTION
**Pull request by ak3n**
This PR extends `api/contract/instances` and `api/contract/wallet/...` endpoints with a GET param `status` that supports filtering of the instances according to the provided status ("active", "stopped", "done"). Returns all instances by defaults.

Also, `ContractInstanceClientState` got a `cicStatus` field with instance's activity status.

Fix #3971

---

Pre-submit checklist:
- Branch
    - [ ] Tests are provided (if possible)
    - [x] Commit sequence broadly makes sense
    - [x] Key commits have useful messages
    - [x] Relevant tickets are mentioned in commit messages
    - [ ] Formatting, materialized Nix files, PNG optimization, etc. are updated
- PR
    - [x] Self-reviewed the diff
    - [x] Useful pull request description
    - [x] Reviewer requested

### CODE REVIEW

File: plutus-contract/src/Wallet/Types.hs [hunk 1]
```diff
@@ -117,6 +119,15 @@ contractInstanceIDs = ContractInstanceId <$> UUID.mockUUIDs
 randomID :: IO ContractInstanceId
 randomID = ContractInstanceId <$> UUID.nextRandom
 
+data ContractActivityStatus = Active | Stopped | Done deriving (Eq, Show, Generic, ToJSON, FromJSON, OpenApi.ToSchema)
```
Comment (silky) [inline plutus-contract/src/Wallet/Types.hs:122]: Could we just re-use the `Activity` type from `ContractInstance.STM` ?

File: plutus-pab/src/Plutus/PAB/Db/Memory/ContractStore.hs [hunk 2]
```diff
@@ -67,7 +67,7 @@ handleContractStore = \case
         case result of
             Just s  -> pure s
             Nothing -> throwError (ContractInstanceNotFound instanceId)
-    Contract.GetActiveContracts -> do
+    Contract.GetContracts _ -> do
```
Comment (silky) [inline plutus-pab/src/Plutus/PAB/Db/Memory/ContractStore.hs:70]: You don't want to use the filter here?

File: plutus-pab/src/Plutus/PAB/Simulator.hs [hunk 3]
```diff
@@ -622,7 +622,7 @@ handleContractStore = \case
         case result of
             Just s  -> pure s
             Nothing -> throwError (ContractInstanceNotFound instanceId)
-    Contract.GetActiveContracts -> do
+    Contract.GetContracts _ -> do
```
Comment (silky) [inline plutus-pab/src/Plutus/PAB/Simulator.hs:625]: Nor here?

File: plutus-pab/src/Plutus/PAB/Webserver/API.hs [hunk 4]
```diff
@@ -43,8 +44,8 @@ type API t walletId -- see note [WalletID type in wallet API]
                         :<|> "stop"     :> Description "Terminate the instance." :> Put '[JSON] ()
                         )
                     )
-            :<|> "instances" :> "wallet" :> Capture "wallet-id" walletId :> Description "List of all active contract instances for the wallet." :>  Get '[JSON] [ContractInstanceClientState t]
-            :<|> "instances" :> Description "List of all active contract instances." :> Get '[JSON] [ContractInstanceClientState t]
+            :<|> "instances" :> "wallet" :> Capture "wallet-id" walletId :> QueryParam "status" Text :> Description "List of contract instances for the wallet filtered by status (active, stopped, done). All by default." :>  Get '[JSON] [ContractInstanceClientState t]
```
Comment (silky) [inline plutus-pab/src/Plutus/PAB/Webserver/API.hs:47]: This looks fine; but is it possible with servant to do this in a type-safe way?

**Review by silky — APPROVED** [review]
Looks good; few comments requesting a bit of clarification!

File: plutus-pab/src/Plutus/PAB/Db/Memory/ContractStore.hs (hunk 2 above)
Comment (ak3n) [reply plutus-pab/src/Plutus/PAB/Db/Memory/ContractStore.hs:70]: That's a good question. I hope @j-mueller will help to answer it.

The thing is that `Contract.PutStopInstance` handler is empty `pure ()` which means that we don't stop instances actually with this type of memory?

File: plutus-pab/src/Plutus/PAB/Webserver/API.hs (hunk 4 above)
Comment (ak3n) [reply plutus-pab/src/Plutus/PAB/Webserver/API.hs:47]: Initially I was going to use `ContractActivityStatus` defined above, but then I remembered about the issue with the purescript generator as it has problems with types there. That's why we use `Capture "contract-instance-id" Text`. Maybe it's only for newtypes and `data` types are fine. Not sure, worth to test.

File: plutus-contract/src/Wallet/Types.hs (hunk 1 above)
Comment (ak3n) [reply plutus-contract/src/Wallet/Types.hs:122]: I believe it's better to separate API and domain types as they tend to change over time and it breaks the API without notices.

File: plutus-pab/src/Plutus/PAB/Db/Memory/ContractStore.hs (hunk 2 above)
Comment (silky) [reply plutus-pab/src/Plutus/PAB/Db/Memory/ContractStore.hs:70]: Yes good point; you're right, it's fine then.

Comment (j-mueller) [reply plutus-pab/src/Plutus/PAB/Db/Memory/ContractStore.hs:70]: Yes, we don't actually track this for the in-memory store (it's only supposed to be a debugging / demo tool)

File: plutus-pab/src/Plutus/PAB/Webserver/API.hs (hunk 4 above)
Comment (j-mueller) [reply plutus-pab/src/Plutus/PAB/Webserver/API.hs:47]: Yes it would be worth testing it, but I am afraid that the purescript bridge won't support it.

**Review by j-mueller — APPROVED** [review]
LGTM

**Pull request merged**